internal/appsec/waf.go

```diff
@@ -95,9 +95,35 @@ func registerWAF(rules []byte, timeout time.Duration, limiter Limiter, obfCfg *O
 // newWAFEventListener returns the WAF event listener to register in order to enable it.
 func newHTTPWAFEventListener(handle *waf.Handle, addresses []string, timeout time.Duration, limiter Limiter) dyngo.EventListener {
 	var monitorRulesOnce sync.Once // per instantiation
+	actionHandler := httpsec.NewActionsHandler()
 
 	return httpsec.OnHandlerOperationStart(func(op *httpsec.Operation, args httpsec.HandlerOperationArgs) {
 		var body interface{}
+		wafCtx := waf.NewContext(handle)
+		if wafCtx == nil {
+			// The WAF event listener got concurrently released
+			return
+		}
+
+		values := map[string]interface{}{}
+		for _, addr := range addresses {
+			if addr == httpClientIP && args.ClientIP.IsValid() {
+				values[httpClientIP] = args.ClientIP.String()
+			}
+		}
+
+		matches, actionIds := runWAF(wafCtx, values, timeout)
+		actionIds = append(actionIds, "block")
+		if len(matches) > 0 {
+			for _, id := range actionIds {
+				actionHandler.Apply(id, op)
+			}
+			op.AddSecurityEvents(matches)
+			log.Debug("appsec: WAF detected an attack before executing the request")
+			if len(actionIds) > 0 {
+				return
+			}
```

Comment:
let's make this even more explicit: please add a `blocked` boolean in the operation to set to true when the action was a blocking one and check it here:
```suggestion
			if op.blocked {
				return
			}
```

Reply:
The operation hasn't necessarily been blocked yet though, we only know that the action is supposed to block it. I think it would be semantically more correct to check the action type, or maybe `ActionHandler.Apply()` could return some status indicating whether applying this action should interrupt the flow here. WDYT?

Reply:
See https://github.com/DataDog/dd-trace-go/pull/1533/commits/784323059db23d711f065ed523f853a3fde57ebb